app/services/fact_cleaner.rb

```diff
@@ -0,0 +1,44 @@
+class FactCleaner
+  delegate :logger, :to => :Rails
+
+  def clean!
+    @to_delete = []
+    logger.info "Searching for orphaned facts..."
+    find_leaf_orphaned_facts
+
+    # we also delete all composes that either don't have any descendant or all of them are to be deleted
+    logger.info "Searching for compose orphaned facts..."
+    find_compose_orphaned_facts
+
+    logger.info "#{@to_delete.size} facts will be removed"
+
+    delete!
+    logger.info "Cleanup of facts finished!"
+    @to_delete.count
+  end
+
+  private
+
+  def find_leaf_orphaned_facts
+    FactName.where(:compose => false).includes(:fact_values).find_in_batches do |batch|
```

Comment:
I don't think `FactName.where(:compose => false).includes(:fact_values)` should be put in a scope? That fact we include fact_values is specific to our local usage. The condition itself could be useful - added `.leaves` and `.composes`.